### ISSUE TITLE
Possible resource leak in IPKPassTemplate implementations

### ISSUE THREAD

**Issue by JocoNagy** [drallgood/jpasskit#117]
The PKInMemorySigningUtil uses the getAllFiles() method from IPKPassTemplate to load the images, however in both implementations (PKPassTemplateFolder and PKPassTemplateInMemory)  it never closes the file input streams.

**Comment by stepio** [comment]
Hello @JocoNagy ,

Thanks for reporting this. I do not own this repo and cannot release new versions, but I'm responsible for a significant amount of changes here, so still feel responsibility for this buddy.

So... talking about your finding, seems that you're right. But as I see, if you use just `provisionPassAtDirectory` you won't be affected by this issue - at least I used this method in my case.

But getting back to what you reported, it's quite easy to fix this in `PKPassTemplateFolder` - just closing the stream inside of the method. But I'm not really sure about `PKPassTemplateInMemory`. Simple approach would be to to close all the streams in `getAllFiles`, but in my opinion this would be conceptually wrong as current instance of `PKPassTemplateInMemory` will not be usable any more. So I'd prefer to make the whole `PKPassTemplateInMemory` implementing `AutoClosable`. This way each instance will stay valid until the relevant `close` is invoked, but now it will be your responsibility.

Any comments, gentlemen?

P.S.: I may do the appropriate PR or can leave to someone else.

**Comment by JocoNagy** [comment]
Hi @stepio,

Thank you for your response.
First I used `PKFileBasedSigningUtil `with `PKPassTemplateFolder `(it uses the `provisionPassAtDirectory`) and  it worked fine. However the pictures I used were on the classpath and when i started my app in a docker container it was not able load the images because they were packed in the Jar file. So i tried the `PKPassTemplateInMemory `with `PKInMemorySigningUtil `and then I found out that the input streams are not closed.

> Simple approach would be to to close all the streams in `getAllFiles`, but in my opinion this would be conceptually wrong as current instance of `PKPassTemplateInMemory` will not be usable any more.

If you want to use `PKPassTemplateInMemory` you need a new instance every time because if `provisionPassAtDirectory` is called it will close the input streams and blows up at the second use, if `getAllFiles` is called it will work fine at the first time but subsequent calls will read 0 byte because you cannot read the streams again.  So closing all the streams will affect the class in the way that `getAllFiles` would blow up just  like `provisionPassAtDirectory` at the second use. This reveals another issue (If this is not by design) because I would expect that if I create a template and add all the images i need i can use this instance any time i want.
I think to fix both problems `Map<String, InputStream>` should be changed for example `Map<String, ByteBuffer>` so the buffers would be created only once and kept in memory. This would also simplify both `provisionPassAtDirectory` and `getAllFiles` methods.

The only working approach for me at the moment is to use `PKPassTemplateInMemory` with `PKFileBasedSigningUtil` but i have to create a new instance of the template every time i want to sign a pass.

Also if PR-s are welcome I would be happy to open one according to my suggestion.

**Comment by stepio** [comment]
Hi again, @JocoNagy, 

> if `getAllFiles` is called it will work fine at the first time but subsequent calls will read 0 byte because you cannot read the streams again

Thanks for reminding this, you're right. So ignore my suggestions above.

I think your idea is perfect:
> think to fix both problems Map<String, InputStream> should be changed for example Map<String, ByteBuffer> so the buffers would be created only once and kept in memory

Such implementation change does not affect the API, so I don't see any negative side-effects in your idea.

> Also if PR-s are welcome I would be happy to open one according to my suggestion.

I cannot guarantee anything as I'm not related to @drallgood . But as per my experience he's cool guy and accepts PRs gratefully. So I'd recommend to implement this change as it would really improve the library.

**Comment by drallgood** [comment]
Fully agree with everything @stepio said. 
Pull requests are always welcome :)
